pkg/controller/rbac/provider/roles/reconciler.go

```diff
@@ -0,0 +1,220 @@
+	return func(r *Reconciler) {
+		r.client = ca
+	}
+}
+
+// WithClusterRoleRenderer specifies how the Reconciler should render RBAC
+// ClusterRoles.
+func WithClusterRoleRenderer(rr ClusterRoleRenderer) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.rbac = rr
+	}
+}
+
+// NewReconciler returns a Reconciler of ProviderRevisions.
+func NewReconciler(mgr manager.Manager, opts ...ReconcilerOption) *Reconciler {
+	kube := unstructured.NewClient(mgr.GetClient())
+
+	r := &Reconciler{
+		mgr: mgr,
+
+		// TODO(negz): Is Updating appropriate here? Probably.
+		client: resource.ClientApplicator{
+			Client:     kube,
+			Applicator: resource.NewAPIUpdatingApplicator(kube),
+		},
+
+		rbac: ClusterRoleRenderFn(RenderClusterRoles),
+
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+	return r
+}
+
+// A Reconciler reconciles ProviderRevisions.
+type Reconciler struct {
+	client resource.ClientApplicator
+	mgr    manager.Manager
+	rbac   ClusterRoleRenderer
+
+	log    logging.Logger
+	record event.Recorder
+}
+
+// Reconcile a ProviderRevision by creating a series of opinionated ClusterRoles
+// that may be bound to allow access to the resources it defines.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	pr := &v1alpha1.ProviderRevision{}
+	if err := r.client.Get(ctx, req.NamespacedName, pr); err != nil {
+		// In case object is not found, most likely the object was deleted and
+		// then disappeared while the event was in the processing queue. We
+		// don't need to take any action in that case.
+		log.Debug(errGetPR, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPR)
+	}
+
+	log = log.WithValues(
+		"uid", pr.GetUID(),
+		"version", pr.GetResourceVersion(),
+		"name", pr.GetName(),
+	)
+
+	if meta.WasDeleted(pr) {
+		// There's nothing to do if our PR is being deleted. Any ClusterRoles
+		// we created will be garbage collected by Kubernetes.
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	l := &v1beta1.CustomResourceDefinitionList{}
+	if err := r.client.List(ctx, l); err != nil {
+		log.Debug(errListCRDs, "error", err)
+		r.record.Event(pr, event.Warning(reasonApplyRoles, errors.Wrap(err, errListCRDs)))
+		return reconcile.Result{RequeueAfter: shortWait}, nil
+	}
+
+	// Filter down to the CRDs that are owned by this ProviderRevision - i.e.
+	// those that it may become the active revision for.
+	crds := make([]v1beta1.CustomResourceDefinition, 0)
+	for _, crd := range l.Items {
+		for _, ref := range crd.GetOwnerReferences() {
+			if ref.UID == pr.GetUID() {
+				crds = append(crds, crd)
+			}
+		}
+	}
+
+	for _, cr := range r.rbac.RenderClusterRoles(pr, crds) {
```

Comment:
I _do_ think we need the cluster roles that are created for users to view managed resource to exist for inactive provider revisions as those CRDs will continue to exist in the cluster even if they are not being reconciled :+1: